### ISSUE TITLE
Overlap between `receptacleObjectIds` and `receptacleParent` in the same object

### ISSUE THREAD

**Issue by gautierdag** [allenai/ai2thor#1088]
There seems to be an inconsistency in some environments where receptacle objects (eg. `Plate`) can get assigned their `receptacleParent` (eg. `CounterTop`) in their `receptacleObjectIds` as well. This implies that `Plate` is both contained by the `CounterTop` and also contains the `CounterTop`. 

Running the following script, I find all instances where this happens.

```python
from ai2thor.controller import Controller

for i in range(1, 31):
    controller = Controller(scene=f"FloorPlan{i}")
    for obj in controller.last_event.metadata["objects"]:
        if obj["receptacleObjectIds"] is not None and obj["parentReceptacles"] is not None:
            intersection = set(obj["receptacleObjectIds"]).intersection(set(obj["parentReceptacles"]))
            if len(intersection) > 0:
                for intersected_object in intersection:
                    print(f"Floor plan {i} has that {obj['objectId']} both contains {intersected_object} and is contained by {intersected_object}")
    controller.stop()
```
```
Floor plan 5 has that Cabinet|-01.15|+02.27|-00.84 both contains Cabinet|-01.15|+02.27|-01.42 and is contained by Cabinet|-01.15|+02.27|-01.42
Floor plan 5 has that Cabinet|+00.12|+02.27|-02.00 both contains Cabinet|-00.46|+02.27|-02.00 and is contained by Cabinet|-00.46|+02.27|-02.00
Floor plan 5 has that Cabinet|-00.46|+02.27|-02.00 both contains Cabinet|+00.12|+02.27|-02.00 and is contained by Cabinet|+00.12|+02.27|-02.00
Floor plan 5 has that CounterTop|+01.16|+00.95|-02.01 both contains Plate|-01.12|+00.90|-01.65 and is contained by Plate|-01.12|+00.90|-01.65
Floor plan 5 has that Cabinet|-01.15|+02.27|-01.42 both contains Cabinet|-01.15|+02.27|-00.84 and is contained by Cabinet|-01.15|+02.27|-00.84
Floor plan 5 has that Plate|-01.12|+00.90|-01.65 both contains CounterTop|+01.16|+00.95|-02.01 and is contained by CounterTop|+01.16|+00.95|-02.01
Floor plan 16 has that Cabinet|+01.88|+00.38|+01.02 both contains Floor|+00.00|+00.00|+00.00 and is contained by Floor|+00.00|+00.00|+00.00
Floor plan 16 has that Cabinet|+02.47|+00.38|+00.38 both contains Floor|+00.00|+00.00|+00.00 and is contained by Floor|+00.00|+00.00|+00.00
Floor plan 16 has that Floor|+00.00|+00.00|+00.00 both contains Cabinet|+02.47|+00.38|+00.38 and is contained by Cabinet|+02.47|+00.38|+00.38
Floor plan 16 has that Floor|+00.00|+00.00|+00.00 both contains Cabinet|+01.88|+00.38|+01.02 and is contained by Cabinet|+01.88|+00.38|+01.02
Floor plan 18 has that CounterTop|+01.39|+01.03|+00.90 both contains Cup|+01.18|+00.98|+03.39 and is contained by Cup|+01.18|+00.98|+03.39
Floor plan 18 has that Cup|+01.18|+00.98|+03.39 both contains CounterTop|+01.39|+01.03|+00.90 and is contained by CounterTop|+01.39|+01.03|+00.90
Floor plan 30 has that CounterTop|+00.13|+00.94|-01.46 both contains Pan|+00.34|+00.87|-01.65 and is contained by Pan|+00.34|+00.87|-01.65
Floor plan 30 has that Pan|+00.34|+00.87|-01.65 both contains CounterTop|+00.13|+00.94|-01.46 and is contained by CounterTop|+00.13|+00.94|-01.46
```

Following the documentation:
```
The object metadata also indicates if a Receptacle object contains any other objects via the receptacleObjectIds metadata string list, which lists all objectIds of any sim objects contained by the Receptacle. Additionally, objects report back any receptacles that they are contained by via the parentReceptacles list of strings in the object metadata.
```

Therefore, I believe expected behavior is for the `parentReceptacles` and `receptacleObjectIds` of an object not to have interesting elements.

Version: `ai2thor=='4.3.0'`

**Comment by winthos** [comment]
Hello! Thank you for bringing this to our attention. I have investigated each of the interactions you listed above, and found some of them were due to inconsistencies in some collider placements that have been corrected with commit [f357daf](https://github.com/allenai/ai2thor/commit/f357dafff4f4ef931a637c4a213672b02e9af8d5), but other aspects of your issue are due to some nuances of how the receptacle system works and does seem to be working as intended.

You note that you don't expect the `parentReceptacles` and `receptacleObjectIds` to have intersecting elements, but there are edge cases where you would see this happen.

The way the receptacle system works is, every receptacle object has some number of box volumes that determine the `receptacle area` for that receptacle sim object. If another sim object's physical collision intersects the `receptacle area` then it is then considered to be contained by that receptacle object. You can visualize this here with a side perspective cutout of a bowl containing this red ball. The bowl's collision is in grey while the bowl's `receptacle area` I've drawn in a green outline box.

![Ball_In_Bowl](https://user-images.githubusercontent.com/17502636/193698179-218d0c18-72d5-4c8c-86d4-4ab28158fed6.png)

However, there are configurations of objects that could lead to two receptacle objects both containing each other, such as if two bowl objects were stacked on top of each other like this. Here, each bowl's physical colliders are intersecting with the other bowl's `receptacle are` and thus, they would both share an intersection in the `parentReceptacles` and `receptacleObjectIds`.

![Bowl_In_Bowl](https://user-images.githubusercontent.com/17502636/193698352-72f27702-1dfe-4514-8b7d-60f0627ad4f4.png)

Finally, there are edge cases where, due to the quirks of how collider volumes detect other colliders, when some bowls or cups with especially low to the "bottom" receptacle zones are placed on a receptacle like a `countertop` or a `floor`, it will show the floor as being contained by the smaller receptacle resting on top of it, as seen here.

![Bowl_On_Floor](https://user-images.githubusercontent.com/17502636/193698498-b03bcc55-f37f-419d-9286-2b7f9252a37c.png)

So while the majority of objects should not have intersections between the `parentReceptacles` and the `receptacleObjectIds`, there are some edge cases, especially when considering multiple receptacle sim objects touching each other, that this can happen.

**Comment by gautierdag** [comment]
Hey @winthos, 
Thanks a lot for the detailed response and diagrams! It makes sense, even though both intuitively and symbolically it's odd to have a Cup contain a CounterTop, but since it's expected behavior I will close this issue.
